### DISCUSSION TITLE
PTC with this core ?!

### DISCUSSION THREAD

**Discussion by coderatchance** [SpenceKonde/megaTinyCore#953] [Library discussion]
Hello all!

One of the things the new Tinys are useful for is also the PTC or the Peripheral Touch Controller due to plenty of GPIOs.
Indeed, the Microchip, etc. offers a dedicated library to handle the PTC, which is pretty well dedicated and hidden as well. 
However, I have not seen any support for teh PTC within this core or am I absolutely blind?
Anyway, has anybody tried to use this core for the PTC as well? 

Best.

**Comment by MX682X** [comment]
I'm planning to release an alpha version of a library in the next few days. Here is a preview, if you want, but it lacks a LOT of "paperwork". All in all it ready to use though, but might contain some bugs. Also, only tested on 1614. [sketch_apr04a-011.zip](https://github.com/SpenceKonde/megaTinyCore/files/11431662/sketch_apr04a-011.zip)

Features: Mutual, selfcap (+ shield) button support (slidery might be user-implementable), suspend the ptc to make adc conversions (not supported in low-power mode), low power mode for an event driven conversion (up to the user to set up event routing), can be used easily outside Arduino/mTC, reduced memory footprint compared to "established solutions"

**Comment by SpenceKonde** [comment]
The main reason we don't have it is that microchip locks it behind their proprietary library (Why? What're we gonna do with that knowledge? Is it going to make us buy more or fewer chips if the touch library is open? I wager "more" so it's kinda illogical. The above version had to be reverse engineered and rebuilt! (like LadaAda did for one line of SAM ARM chips)

@MX682X  - you're planning to release as separate library? Great let me know when you do and i will link to it from supported libraries. I was not looking forward to trying to write the docs I'd need in order to include that library

**Reply by MX682X** [reply]
Yeah, I'm planning to publish this in a seperate repository, for now at least. Ultimately it would be best to add it to the list of included libraries like Wire, as there aren't any other Arduino board sets that would benefit of it. The problem is though, that at this point, the library has to be considered experimental and is prone to changes and bug fixes. It would be a big hassle to open Pull-Requests and so on to fix it here. It would also take significantly more time until a fix would be released, so I concluded to keep it on my end at first. Might be still cool to add a link on the main page to advertise it to gather feedback, this might lead to changes of the API, (hence experimental)

Also, I was never expeting of you to write the docs for it, maybe proofread and annotate unclear passages, but nothing more then that.

On a side note: It is possible to set up an Arduino library to link against a pre-compiled .a file, that qTouch comes in, but I failed to do so (I've discovered https://github.com/arduino-libraries/Arduino_MCHPTouch only later), but looking at the assembly listing and the way gcc was handling `for (uint8_t i = 0; i++; i<size) someArrayOfStructs[i];` (uses one reg for i, two regs for base pointer, two regs for base pointer offset, two regs for the pointer for the actual access) I was like: nope, that can be done better (with a single-linked list, removing the need for pointer arithmetics. (The library doesn't make any pointer arithmetics infact (as far as I'm aware), being even better then qTouch in that regard, but it doesn't matter much) (all pointers in the list are even guranteed to be in ram by a simple *highbyte* < 0x38 signed comparison when creating the list))

**Reply by SpenceKonde** [reply]
Cool Coincidentally, it looks like theres's going to be some action on the PTC for the DA-series too for Arduino.

**Comment by SpenceKonde** [comment]
Closing as does not represent an action item for me.
